### ISSUE TITLE
add CLI command to delete apps on your tenant

### ISSUE THREAD

**Issue by jstrachan** [fabric8io/gofabric8#400]
see this issue for background:
https://github.com/fabric8io/fabric8-init-tenant/issues/75

it might be nice for folks be able to do something like

```
gofabric8 delete apps --all

gofabric8 delete apps foo bar
```

which would delete 

* the BuildConfig / Build resources matching (foo|bar) or "*") in the ${ID} namespace
* remove the matching resources in `${ID}-test, ${ID}-stage, ${ID}-run` namespaces

**Comment by jstrachan** [comment]
as a first cut we can detect `ID` by using `oc whoami`. Or we could do the trick that the `gofabric8 che` command does and look at all projects and look for `*-che` or `*-jenkins` too ;)
https://github.com/fabric8io/gofabric8/blob/master/cmds/che.go#L80-L85

Also rather than hard coding `${ID}-run` / `${ID}-test` / `${ID}-stage` we could reuse the code to load the environments to find those namespaces via the `fabric8-environments ConfigMap` in the `{ID}` namespace 

We should log a warning if we don't find a `fabric8-environments ConfigMap` in the `ID` namespace - its maybe a sign that we've found the wrong `ID` namespace. We can also let users specify the `ID` namespace on the CLI too. 

If we don't find the `fabric8-environments ConfigMap` we maybe should assume `${ID}-run` / `${ID}-test` / `${ID}-stage` (if those projects exist) - as the user might have accidentally zapped the ConfigMap?

**Comment by jstrachan** [comment]
the delete all apps command would look like this:
```
oc delete bc --all -n ${ID}
oc delete build --all -n ${ID}
oc delete all --all -n ${ID}-test
oc delete all --all -n ${ID}-stage
oc delete all --all -n ${ID}-run
```

then obviously we replace the `--all` when deleting individual things.  

For an app `foo` we'll have 
* a BC called `foo` in namespace `${ID}`
* Builds with labels `buildconfig: foo` in namespace `${ID}`

Then for each environment (e.g. `${ID}-test, ${ID}-stage, ${ID}-run`) we'll need to delete 
* Deployment
* DeploymentConfigs
* ReplicaSets / ReplicationControllers (should be deleted by above really)
* Pods (should be deleted by above really)
* Services
* Routes
* PVCs
* Secrets & ConfigMaps

To find the resources to delete we could use these rules:

* if the name is `foo`
* if the resource has a label `project: foo` or `app:foo`

Right now we tend to use `project: foo` for fabric8-maven-plugin generated resources; though the `oc` approach is to use `app` label which we'll probably migrate to soon https://github.com/fabric8io/fabric8-maven-plugin/issues/945

**Comment by jstrachan** [comment]
Actually on the last point; we may want to reuse that logic (given a namespace, delete one or more apps resources by name) as a generally useful openshift/kubernetes command.

So maybe we need 2 separate CLI commands; one to delete buildconfigs+ builds + apps across all environments and one to just delete apps within the current or a specific namespace?

Not quite sure how best to name these 2 things though.  Anyone got any good ideas?

* delete an app in a local namespace: `gofabric8 delete app foo`
* delete all app resources across all environments: `gofabric8 delete app-environments foo` or for short `gofabric8 delete app-env foo`

**Comment by chmouel** [comment]
About cleanup all, isn't that handle by the cleanup command (https://github.com/fabric8io/gofabric8/blob/master/cmds/cleanup.go#L38-L39) or is there some differences between them? which could be confusing for the user

**Comment by jstrachan** [comment]
@chmouel great catch yeah! The cleanup command pre-dates the new init-tenant model of openshift.io; so tries to just zap the main apps installed rather than dealing with some/all apps in the user tenant stuff.

I guess its confusing though; so wonder if we should maybe have separate commands like; `gofabric8 delete app` versus `gofabric8 delete system`? (The latter being the rename of cleanup)?

**Comment by chmouel** [comment]
> I guess its confusing though; so wonder if we should maybe have separate commands like; gofabric8 delete app versus gofabric8 delete system? (The latter being the rename of cleanup)?

yep it is very much make sense! sounds like a plan batman!

**Comment by chmouel** [comment]
As for the delete app, the logic would be implementing in gofabric8 right ? 

(ID would be detected with the logic from [cmds/common.go:detectCurrentUserProject ](https://github.com/fabric8io/gofabric8/blob/master/cmds/common.go#L80-L81) and apps would be given by the user on the command line args)

```zsh
APP=gogs
ID=developer

oc delete all -n ${ID} -l project=${APP}
for i (che jenkins run stage) oc delete all -n $ID-${i} -l project=${APP}
```

**Comment by jstrachan** [comment]
yeah that logic looks right; where the list of namespaces comes from the `fabric8-environments` in the core users namespace plus `$ID-che` and $ID-jenkins`

**Comment by jstrachan** [comment]
we should check for both`project=`${APP}` and `app=${APP}` labels as at some point we'll switch to use the latter to be consistent with `oc new-app`https://github.com/fabric8io/fabric8-maven-plugin/issues/945

**Comment by jstrachan** [comment]
actually one other nit - I've noticed `ConfigMap`'s don't usually get deleted by `oc delete all`

**Comment by chmouel** [comment]
yep, sorry that PR didn't address the whole things, seems like the PVC too are not getting deleted so will add that too 👍

**Comment by chmouel** [comment]
I believe the configmaps wasn't getting deleted although written :  https://github.com/fabric8io/gofabric8/blob/master/cmds/cleanup.go#L303-L315

is because we are trying to delete ingress routes just earlier and it fails due of the wrong right (as per screenshots)

I am going to change the ones who needs admin rights access to a warning, unless you see a better solution ?

![image](https://user-images.githubusercontent.com/98980/27734330-ad5d1e00-5d9a-11e7-82ba-b825d0c3a6bd.png)

**Comment by chmouel** [comment]
So following the discussion we had directly, here is the recap of the different implemented type, let me know if I get this rights : 

* oc delete system : 

    delete all pvc first 
    delete all projects on openshift
    delete all namespace on kubernetes

* oc delete app APP:

   delete all resource from APP including bc, secrets etc..

* oc delete app --all

  delete all apps from the current TENANT
  (no project/ns deletion)

* oc delete tenant TENANT

  delete all auto generated stuff for the tenant like all the tenant generated stuff like the run/stage namespaces + the jenkins + che namespace 
 (are we including the projects deletion as well in there?)

Let me know if that looks good for you, Thanks,

**Comment by jstrachan** [comment]
looks good - be sure when removing the tenant namespaces to delete all the PVCs first too, to be sure

**Comment by jstrachan** [comment]
also note that deleting a tenant is different from deleting the apps in a tenant. i.e. deleting all the apps should keep the namespaces around & not delete jenkins or che from $TENANT-jenkins or $TENANT-che.

So if someone wants to clean their tenant, they'd do `gofabric8 delete app --all` whereas to uninstall their tenant completely there's `gofabric8 delete tenant foo`

**Comment by chmouel** [comment]
cool thanks! as for deleting the pvc belonging to a certain tenant, how do we detect that pvc belonged to that tenant?

**Comment by jstrachan** [comment]
deleting the system, deletes the PVCs and namespace of the system namespace

deleting a tenant deletes the PVCs and the namespace for each tenant namespace (e.g. `$TENANT, $TENANT-che,  $TENANT-jenkins,  $TENANT-stage,  $TENANT-run`)

deleting an app `foo` removes PVCs and then resources which match the label `project: foo` or `app: foo` in the tenant namespaces (mostly just `$TENANT, $TENANT-stage,  $TENANT-run`)
